File: app/src/main/java/com/readrops/app/database/dao/FeedDao.java
package com.readrops.app.database.dao;


import androidx.lifecycle.LiveData;
import androidx.room.Dao;
import androidx.room.Query;
import androidx.room.Transaction;

import com.readrops.app.database.entities.Account;
import com.readrops.app.database.entities.Feed;
import com.readrops.app.database.pojo.FeedWithFolder;

import java.util.ArrayList;
import java.util.List;

@Dao
public abstract class FeedDao implements BaseDao<Feed> {

    @Query("Select * from Feed Where account_id = :accountId order by name ASC")
    public abstract List<Feed> getAllFeeds(int accountId);

    @Query("Delete From Feed Where id = :feedId")
    public abstract void delete(int feedId);

    @Query("Select case When :feedUrl In (Select url from Feed Where account_id = :accountId) Then 1 else 0 end")
    public abstract boolean feedExists(String feedUrl, int accountId);

    @Query("Select case When :remoteId In (Select remoteId from Feed Where account_id = :accountId) Then 1 else 0 end")
    public abstract boolean remoteFeedExists(String remoteId, int accountId);

    @Query("Select count(*) from Feed Where account_id = :accountId")
    public abstract int getFeedCount(int accountId);

    @Query("Select * from Feed Where url = :feedUrl")
    public abstract Feed getFeedByUrl(String feedUrl);

    @Query("Select id from Feed Where remoteId = :remoteId And account_id = :accountId")
    public abstract int getFeedIdByRemoteId(String remoteId, int accountId);

    @Query("Select * from Feed Where folder_id = :folderId")
    public abstract List<Feed> getFeedsByFolder(int folderId);

    @Query("Select * from Feed Where account_id = :accountId And folder_id is null")
    public abstract List<Feed> getFeedsWithoutFolder(int accountId);

    @Query("Update Feed set etag = :etag, last_modified = :lastModified Where id = :feedId")
    public abstract void updateHeaders(String etag, String lastModified, int feedId);

    @Query("Update Feed set name = :feedName, url = :feedUrl, folder_id = :folderId Where id = :feedId")
    public abstract void updateFeedFields(int feedId, String feedName, String feedUrl, Integer folderId);

    @Query("Update Feed set name = :name, folder_id = :folderId Where remoteId = :remoteFeedId And account_id = :accountId")
    public abstract void updateNameAndFolder(String remoteFeedId, int accountId, String name, Integer folderId);

    @Query("Update Feed set text_color = :textColor, background_color = :bgColor Where id = :feedId")
    public abstract void updateColors(int feedId, int textColor, int bgColor);

    @Query("Select Feed.name as feed_name, Feed.id as feed_id, Folder.name as folder_name, Folder.id as folder_id, Folder.remoteId as folder_remoteId," +
            "Feed.description as feed_description, Feed.icon_url as feed_icon_url, Feed.url as feed_url, Feed.folder_id as feed_folder_id" +
            ", Feed.siteUrl as feed_siteUrl, Feed.remoteId as feed_remoteId from Feed Left Join Folder on Feed.folder_id = Folder.id Where Feed.account_id = :accountId Order by Feed.name")
    public abstract LiveData<List<FeedWithFolder>> getAllFeedsWithFolder(int accountId);

    @Query("Select * From Feed Where id in (:ids)")
    public abstract List<Feed> selectFromIdList(List<Long> ids);

    @Query("Select remoteId From Feed Where account_id = :accountId")
    public abstract List<String> getFeedRemoteIdsOfAccount(int accountId);

    @Query("Delete from Feed Where id in (:ids)")
    abstract void deleteByIds(List<String> ids);

    @Query("Select id From Folder Where remoteId = :remoteId And account_id = :accountId")
    abstract int getRemoteFolderLocalId(String remoteId, int accountId);

    /**
     * Insert, update and delete feeds, by account
     *
     * @param feeds   feeds to insert or update
     * @param account owner of the feeds
     * @return the list of the inserted feeds ids
     */
    @Transaction
    public List<Long> feedsUpsert(List<Feed> feeds, Account account) {
        List<String> accountFeedIds = getFeedRemoteIdsOfAccount(account.getId());
        List<Feed> feedsToInsert = new ArrayList<>();

        for (Feed feed : feeds) {
            Integer folderId;

            try {
                int remoteFolderId = Integer.parseInt(feed.getRemoteFolderId());
                folderId = remoteFolderId == 0 ? null : getRemoteFolderLocalId(feed.getRemoteFolderId(), account.getId());
            } catch (Exception e) {
                folderId = feed.getRemoteFolderId() == null ? null : getRemoteFolderLocalId(feed.getRemoteFolderId(), account.getId());
            }

            if (remoteFeedExists(feed.getRemoteId(), account.getId())) {
                updateNameAndFolder(feed.getRemoteId(), account.getId(), feed.getName(), folderId);

                accountFeedIds.remove(feed.getRemoteId());
            } else {
                feed.setFolderId(folderId);

                feedsToInsert.add(feed);
            }
        }

        if (!accountFeedIds.isEmpty())
            deleteByIds(accountFeedIds);

        return insert(feedsToInsert);
    }
}

File: db/src/main/java/com/readrops/db/dao/FeedDao.kt
package com.readrops.db.dao

import androidx.room.Dao
import androidx.room.Query
import androidx.room.RawQuery
import androidx.room.Transaction
import androidx.sqlite.db.SupportSQLiteQuery
import com.readrops.db.entities.Feed
import com.readrops.db.entities.Item
import com.readrops.db.entities.OpenIn
import com.readrops.db.entities.account.Account
import com.readrops.db.pojo.FeedWithCount
import com.readrops.db.pojo.FeedWithFolder
import kotlinx.coroutines.flow.Flow

@Dao
interface FeedDao : BaseDao<Feed> {

    @Query("Select * From Feed Where id = :feedId")
    suspend fun selectFeed(feedId: Int): Feed

    @Query("Select * From Feed Where folder_id = :folderId")
    suspend fun selectFeedsByFolder(folderId: Int): List<Feed>

    @Query("Select * from Feed Where account_id = :accountId order by name ASC")
    suspend fun selectFeeds(accountId: Int): List<Feed>

    @Query("Update Feed set etag = :etag, last_modified = :lastModified Where id = :feedId")
    suspend fun updateHeaders(etag: String, lastModified: String, feedId: Int)

    @Query("Select case When :feedUrl In (Select url from Feed Where account_id = :accountId) Then 1 else 0 end")
    suspend fun feedExists(feedUrl: String, accountId: Int): Boolean

    @RawQuery(observedEntities = [Feed::class, Item::class])
    fun selectFeedsWithoutFolder(query: SupportSQLiteQuery): Flow<List<FeedWithCount>>

    @Query("Update Feed set name = :feedName, url = :feedUrl, folder_id = :folderId Where id = :feedId")
    fun updateFeedFields(feedId: Int, feedName: String, feedUrl: String, folderId: Int?)

    @Query("Select count(*) from Feed Where account_id = :accountId")
    suspend fun selectFeedCount(accountId: Int): Int

    @Query("Select remote_id From Feed Where account_id = :accountId")
    suspend fun selectFeedRemoteIds(accountId: Int): List<String>

    @Query("Select id From Folder Where remoteId = :remoteId And account_id = :accountId")
    suspend fun selectRemoteFolderLocalId(remoteId: String, accountId: Int): Int

    @Query("Select id From Feed Where remote_id = :remoteId And account_id = :accountId")
    suspend fun selectRemoteFeedLocalId(remoteId: String, accountId: Int): Int

    @Query("Update Feed set name = :name, folder_id = :folderId Where remote_id = :remoteFeedId And account_id = :accountId")
    suspend fun updateFeedNameAndFolder(remoteFeedId: String, accountId: Int, name: String, folderId: Int?)

    @Query("Delete from Feed Where remote_id in (:ids) And account_id = :accountId")
    suspend fun deleteByIds(ids: List<String>, accountId: Int)

    @Query("Update Feed set color = :color Where id = :feedId")
    suspend fun updateFeedColor(feedId: Int, color: Int)

    @Query("""Select Feed.*, Folder.name as folder_name From Feed Left Join Folder On Feed.folder_id = Folder.id 
        Where Feed.account_id = :accountId Order By Feed.name, Folder.name""")
    fun selectFeedsWithFolderName(accountId: Int): Flow<List<FeedWithFolder>>

    @Query("Update Feed set notification_enabled = :enabled Where id = :feedId")
    suspend fun updateFeedNotificationState(feedId: Int, enabled: Boolean)

    @Query("Update Feed set notification_enabled = :enabled Where account_id = :accountId")
    suspend fun updateAllFeedsNotificationState(accountId: Int, enabled: Boolean)

    @Query("Select * From Feed Where id in (:ids)")
    suspend fun selectFromIds(ids: List<Int>): List<Feed>

    @Query("Update Feed set icon_url = :iconUrl Where id = :feedId")
    suspend fun updateFeedIconUrl(feedId: Int, iconUrl: String)

    @Query("Update Feed set open_in = :openIn Where id = :feedId")
    suspend fun updateOpenInSetting(feedId: Int, openIn: OpenIn)

    @Query("Update Feed set open_in_ask = :openInAsk Where id = :feedId")
    suspend fun updateOpenInAsk(feedId: Int, openInAsk: Boolean)

    /**
     * Insert, update and delete feeds by account
     *
     * This method must always be called with the full [feeds] list
     *
     * @param feeds   feeds to insert or update
     * @param account owner of the feeds
     * @return newly inserted feeds
     */
    @Transaction
    suspend fun upsertFeeds(feeds: List<Feed>, account: Account): List<Feed> {
        val localFeedIds = selectFeedRemoteIds(account.id)

        val feedsToInsert = feeds.filter { feed -> localFeedIds.none { localFeedId -> feed.remoteId == localFeedId } }
        val feedsToDelete = localFeedIds.filter { localFeedId -> feeds.none { feed -> localFeedId == feed.remoteId } }

        feeds.forEach { feed ->
            feed.folderId = if (feed.remoteFolderId == null) {
                null
            } else {
                selectRemoteFolderLocalId(feed.remoteFolderId!!, account.id)
            }

            // works only for already existing feeds
            updateFeedNameAndFolder(feed.remoteId!!, account.id, feed.name!!, feed.folderId)
        }

        if (feedsToDelete.isNotEmpty()) {
            deleteByIds(feedsToDelete, account.id)
        }

        if (feedsToInsert.isNotEmpty()) {
            insert(feedsToInsert)
                .zip(feedsToInsert)
                .forEach { (id, feed) -> feed.id = id.toInt() }

        }

        return feedsToInsert
    }
}